File: scripts/python/test_server_api.py
#!/usr/bin/env python3
"""
Kaetram Game Server API Test Script
"""
import os
import sys
import time
import json
from typing import Dict, List, Optional, Any

# Add utils path
sys.path.insert(0, os.path.join(os.path.dirname(__file__), '..', 'utils'))

from test_utils import TestConfig, TestSuite, validate_response, cleanup_test_agents, generate_test_username

class ServerAPITests:
    def __init__(self, config: TestConfig):
        self.config = config
        self.suite = TestSuite("ServerAPI", config)
        self.base_url = config.get('server.base_url')
        self.test_tokens = []  # Save test tokens for cleanup
        
    def test_server_info(self) -> Dict[str, Any]:
        """Test server information endpoint"""
        response = self.suite.client.get(f"{self.base_url}/")
        data = validate_response(response)
        
        # Validate returned fields
        required_fields = ['name', 'port', 'gameVersion', 'maxPlayers', 'playerCount']
        for field in required_fields:
            if field not in data:
                raise ValueError(f"Missing required field: {field}")
        
        return data
    
    def test_ai_create_agent(self) -> Dict[str, Any]:
        """Test creating AI agent"""
        username = generate_test_username()
        password = "testpass123"
        
        response = self.suite.client.post(f"{self.base_url}/ai/create", json={
            "username": username,
            "password": password
        })
        data = validate_response(response)
        
        # Validate response
        if data.get('status') != 'success':
            raise ValueError(f"Agent creation failed: {data.get('message', 'Unknown error')}")
        
        if 'token' not in data:
            raise ValueError("Missing token field in response")
        
        # Save token for cleanup
        self.test_tokens.append(data['token'])
        
        return {"username": username, "password": password, "token": data['token']}
    
    def test_ai_login(self) -> Dict[str, Any]:
        """Test AI agent login"""
        # First create an agent
        agent_data = self.test_ai_create_agent()
        
        response = self.suite.client.post(f"{self.base_url}/ai/login", json={
            "username": agent_data['username'],
            "password": agent_data['password']
        })
        data = validate_response(response)
        
        # Validate response
        if data.get('status') != 'success':
            raise ValueError(f"Login failed: {data.get('message', 'Unknown error')}")
        
        if 'token' not in data:
            raise ValueError("Missing token field in response")
        
        # Update token
        self.test_tokens.append(data['token'])
        
        return {"token": data['token']}
    
    def test_ai_logout(self) -> Dict[str, Any]:
        """Test AI agent logout"""
        # First login
        login_data = self.test_ai_login()
        token = login_data['token']
        
        response = self.suite.client.post(f"{self.base_url}/ai/logout", json={
            "token": token
        })
        data = validate_response(response)
        
        # Validate response
        if data.get('status') != 'success':
            raise ValueError(f"Logout failed: {data.get('message', 'Unknown error')}")
        
        # Remove token from cleanup list
        if token in self.test_tokens:
            self.test_tokens.remove(token)
        
        return data
    
    def test_ai_move(self) -> Dict[str, Any]:
        """Test AI agent movement"""
        # First login
        login_data = self.test_ai_login()
        token = login_data['token']
        
        # Test movement
        response = self.suite.client.post(f"{self.base_url}/ai/move", json={
            "token": token,
            "x": 100,
            "y": 100
        })
        data = validate_response(response)
        
        # Validate response
        if data.get('status') != 'success':
            raise ValueError(f"Movement failed: {data.get('message', 'Unknown error')}")
        
        return data
    
    def test_ai_stop(self) -> Dict[str, Any]:
        """Test AI agent stop"""
        # First login
        login_data = self.test_ai_login()
        token = login_data['token']
        
        # Test stop
        response = self.suite.client.post(f"{self.base_url}/ai/stop", json={
            "token": token
        })
        data = validate_response(response)
        
        # Validate response
        if data.get('status') != 'success':
            raise ValueError(f"Stop failed: {data.get('message', 'Unknown error')}")
        
        return data
    
    def test_ai_chat(self) -> Dict[str, Any]:
        """Test AI agent chat"""
        # First login
        login_data = self.test_ai_login()
        token = login_data['token']
        
        # Test chat
        response = self.suite.client.post(f"{self.base_url}/ai/chat", json={
            "token": token,
            "message": "Hello, this is a test message!"
        })
        data = validate_response(response)
        
        # Validate response
        if data.get('status') != 'success':
            raise ValueError(f"Chat failed: {data.get('message', 'Unknown error')}")
        
        return data
    
    def test_ai_observe(self) -> Dict[str, Any]:
        """Test AI agent observation"""
        # First login
        login_data = self.test_ai_login()
        token = login_data['token']
        
        # Test observe
        response = self.suite.client.get(f"{self.base_url}/ai/observe", params={
            "token": token
        })
        data = validate_response(response)
        
        # Validate response structure
        expected_fields = ['location', 'map', 'entities', 'inventory', 'player']
        for field in expected_fields:
            if field not in data:
                raise ValueError(f"Missing expected field in observation: {field}")
        
        return data
    
    def test_ai_attack(self) -> Dict[str, Any]:
        """Test AI agent attack"""
        # First login
        login_data = self.test_ai_login()
        token = login_data['token']
        
        # Test attack (using a test target ID)
        response = self.suite.client.post(f"{self.base_url}/ai/attack", json={
            "token": token,
            "target": "test_target_id"
        })
        data = validate_response(response)
        
        # Attack may fail due to no target, but should still return valid response
        if data.get('status') not in ['success', 'error']:
            raise ValueError(f"Invalid attack response: {data}")
        
        return data
    
    def test_ai_equip(self) -> Dict[str, Any]:
        """Test AI agent equipment"""
        # First login
        login_data = self.test_ai_login()
        token = login_data['token']
        
        # Test equip (using a test item index)
        response = self.suite.client.post(f"{self.base_url}/ai/equip", json={
            "token": token,
            "index": 0
        })
        data = validate_response(response)
        
        # Equip may fail due to no item, but should still return valid response
        if data.get('status') not in ['success', 'error']:
            raise ValueError(f"Invalid equip response: {data}")
        
        return data
    
    def test_ai_collect(self) -> Dict[str, Any]:
        """Test AI agent resource collection"""
        # First login
        login_data = self.test_ai_login()
        token = login_data['token']
        
        # Test collect (using a test resource ID)
        response = self.suite.client.post(f"{self.base_url}/ai/collect", json={
            "token": token,
            "resource": "test_resource_id"
        })
        data = validate_response(response)
        
        # Collect may fail due to no resource, but should still return valid response
        if data.get('status') not in ['success', 'error']:
            raise ValueError(f"Invalid collect response: {data}")
        
        return data
    
    def test_ai_craft(self) -> Dict[str, Any]:
        """Test AI agent crafting"""
        # First login
        login_data = self.test_ai_login()
        token = login_data['token']
        
        # Test craft (using a test item key)
        response = self.suite.client.post(f"{self.base_url}/ai/craft", json={
            "token": token,
            "item": "test_item_key"
        })
        data = validate_response(response)
        
        # Craft may fail due to no materials, but should still return valid response
        if data.get('status') not in ['success', 'error']:
            raise ValueError(f"Invalid craft response: {data}")
        
        return data
    
    def test_ai_enter(self) -> Dict[str, Any]:
        """Test AI agent entering portals"""
        # First login
        login_data = self.test_ai_login()
        token = login_data['token']
        
        # Test enter (portal/warp)
        response = self.suite.client.post(f"{self.base_url}/ai/enter", json={
            "token": token
        })
        data = validate_response(response)
        
        # Enter may fail due to no portal, but should still return valid response
        if data.get('status') not in ['success', 'error']:
            raise ValueError(f"Invalid enter response: {data}")
        
        return data
    
    def run_all_tests(self):
        """Run all tests"""
        self.suite.start_time = time.time()
        
        # Define test methods
        test_methods = [
            ("server_info", self.test_server_info),
            ("ai_create_agent", self.test_ai_create_agent),
            ("ai_login", self.test_ai_login),
            ("ai_logout", self.test_ai_logout),
            ("ai_move", self.test_ai_move),
            ("ai_stop", self.test_ai_stop),
            ("ai_chat", self.test_ai_chat),
            ("ai_observe", self.test_ai_observe),
            ("ai_attack", self.test_ai_attack),
            ("ai_equip", self.test_ai_equip),
            ("ai_collect", self.test_ai_collect),
            ("ai_craft", self.test_ai_craft),
            ("ai_enter", self.test_ai_enter),
        ]
        
        # Run each test
        for test_name, test_func in test_methods:
            try:
                self.suite.run_test(test_name, test_func)
            except Exception as e:
                self.suite.logger.error(f"Error running test {test_name}: {e}")
        
        self.suite.end_time = time.time()
        
        # Cleanup
        if self.test_tokens:
            cleanup_test_agents(self.config, self.test_tokens)
        
        # Generate report
        self.suite.save_report()
        self.suite.print_summary()

def main():
    """Main function"""
    try:
        config = TestConfig()
        tests = ServerAPITests(config)
        tests.run_all_tests()
    except Exception as e:
        print(f"Error: {e}")
        sys.exit(1)

if __name__ == "__main__":
    main() 

File: scripts/javascript/test_server_api.js
#!/usr/bin/env node
/**
 * Kaetram Game Server API Test Script - JavaScript Version
 */
const path = require('path');
const { TestConfig, TestSuite, validateResponse, cleanupTestAgents, generateTestUsername } = require('../utils/test_utils');

class ServerAPITests {
    constructor(config) {
        this.config = config;
        this.suite = new TestSuite('ServerAPI', config);
        this.baseUrl = config.get('server.base_url');
        this.testTokens = [];
    }

    async testServerInfo() {
        const response = await this.suite.client.get(`${this.baseUrl}/`);
        const data = validateResponse(response);

        // Validate returned fields
        const requiredFields = ['name', 'port', 'gameVersion', 'maxPlayers', 'playerCount'];
        for (const field of requiredFields) {
            if (!(field in data)) {
                throw new Error(`Missing required field: ${field}`);
            }
        }

        return data;
    }

    async testAICreateAgent() {
        const username = generateTestUsername();
        const password = 'testpass123';

        const response = await this.suite.client.post(`${this.baseUrl}/ai/create`, {
            data: { username, password }
        });
        const data = validateResponse(response);

        if (data.status !== 'success') {
            throw new Error(`Agent creation failed: ${data.message || 'Unknown error'}`);
        }

        if (!data.token) {
            throw new Error('Missing token field in response');
        }

        // Save token for cleanup
        this.testTokens.push(data.token);

        return { username, password, token: data.token };
    }

    async testAILogin() {
        // First create an agent
        const agentData = await this.testAICreateAgent();

        const response = await this.suite.client.post(`${this.baseUrl}/ai/login`, {
            data: {
                username: agentData.username,
                password: agentData.password
            }
        });
        const data = validateResponse(response);

        if (data.status !== 'success') {
            throw new Error(`Login failed: ${data.message || 'Unknown error'}`);
        }

        if (!data.token) {
            throw new Error('Missing token field in response');
        }

        // Update token
        this.testTokens.push(data.token);

        return { token: data.token };
    }

    async testAILogout() {
        // First login
        const loginData = await this.testAILogin();
        const token = loginData.token;

        const response = await this.suite.client.post(`${this.baseUrl}/ai/logout`, {
            data: { token }
        });
        const data = validateResponse(response);

        if (data.status !== 'success') {
            throw new Error(`Logout failed: ${data.message || 'Unknown error'}`);
        }

        // Remove token from cleanup list
        const tokenIndex = this.testTokens.indexOf(token);
        if (tokenIndex > -1) {
            this.testTokens.splice(tokenIndex, 1);
        }

        return data;
    }

    async testAIMove() {
        // First login
        const loginData = await this.testAILogin();
        const token = loginData.token;

        // Test movement
        const response = await this.suite.client.post(`${this.baseUrl}/ai/move`, {
            data: {
                token,
                x: 100,
                y: 100
            }
        });
        const data = validateResponse(response);

        if (data.status !== 'success') {
            throw new Error(`Movement failed: ${data.message || 'Unknown error'}`);
        }

        return data;
    }

    async testAIStop() {
        // First login
        const loginData = await this.testAILogin();
        const token = loginData.token;

        // Test stop
        const response = await this.suite.client.post(`${this.baseUrl}/ai/stop`, {
            data: { token }
        });
        const data = validateResponse(response);

        if (data.status !== 'success') {
            throw new Error(`Stop failed: ${data.message || 'Unknown error'}`);
        }

        return data;
    }

    async testAIChat() {
        // First login
        const loginData = await this.testAILogin();
        const token = loginData.token;

        // Test chat
        const response = await this.suite.client.post(`${this.baseUrl}/ai/chat`, {
            data: {
                token,
                message: 'Hello, this is a test message!'
            }
        });
        const data = validateResponse(response);

        if (data.status !== 'success') {
            throw new Error(`Chat failed: ${data.message || 'Unknown error'}`);
        }

        return data;
    }

    async testAIObserve() {
        // First login
        const loginData = await this.testAILogin();
        const token = loginData.token;

        // Test observe
        const response = await this.suite.client.get(`${this.baseUrl}/ai/observe`, {
            params: { token }
        });
        const data = validateResponse(response);

        // Validate response structure
        const expectedFields = ['location', 'map', 'entities', 'inventory', 'player'];
        for (const field of expectedFields) {
            if (!(field in data)) {
                throw new Error(`Missing expected field in observation: ${field}`);
            }
        }

        return data;
    }

    async testAIAttack() {
        // First login
        const loginData = await this.testAILogin();
        const token = loginData.token;

        // Test attack (using a test target ID)
        const response = await this.suite.client.post(`${this.baseUrl}/ai/attack`, {
            data: {
                token,
                target: 'test_target_id'
            }
        });
        const data = validateResponse(response);

        // Attack may fail due to no target, but should still return valid response
        if (!['success', 'error'].includes(data.status)) {
            throw new Error(`Invalid attack response: ${JSON.stringify(data)}`);
        }

        return data;
    }

    async testAIEquip() {
        // First login
        const loginData = await this.testAILogin();
        const token = loginData.token;

        // Test equip (using a test item index)
        const response = await this.suite.client.post(`${this.baseUrl}/ai/equip`, {
            data: {
                token,
                index: 0
            }
        });
        const data = validateResponse(response);

        // Equip may fail due to no item, but should still return valid response
        if (!['success', 'error'].includes(data.status)) {
            throw new Error(`Invalid equip response: ${JSON.stringify(data)}`);
        }

        return data;
    }

    async testAICollect() {
        // First login
        const loginData = await this.testAILogin();
        const token = loginData.token;

        // Test collect (using a test resource ID)
        const response = await this.suite.client.post(`${this.baseUrl}/ai/collect`, {
            data: {
                token,
                resource: 'test_resource_id'
            }
        });
        const data = validateResponse(response);

        // Collect may fail due to no resource, but should still return valid response
        if (!['success', 'error'].includes(data.status)) {
            throw new Error(`Invalid collect response: ${JSON.stringify(data)}`);
        }

        return data;
    }

    async testAICraft() {
        // First login
        const loginData = await this.testAILogin();
        const token = loginData.token;

        // Test craft (using a test item key)
        const response = await this.suite.client.post(`${this.baseUrl}/ai/craft`, {
            data: {
                token,
                item: 'test_item_key'
            }
        });
        const data = validateResponse(response);

        // Craft may fail due to no materials, but should still return valid response
        if (!['success', 'error'].includes(data.status)) {
            throw new Error(`Invalid craft response: ${JSON.stringify(data)}`);
        }

        return data;
    }

    async testAIEnter() {
        // First login
        const loginData = await this.testAILogin();
        const token = loginData.token;

        // Test enter (portal/warp)
        const response = await this.suite.client.post(`${this.baseUrl}/ai/enter`, {
            data: { token }
        });
        const data = validateResponse(response);

        // Enter may fail due to no portal, but should still return valid response
        if (!['success', 'error'].includes(data.status)) {
            throw new Error(`Invalid enter response: ${JSON.stringify(data)}`);
        }

        return data;
    }

    async runAllTests() {
        this.suite.startTime = Date.now();

        // Define test methods
        const testMethods = [
            ['server_info', () => this.testServerInfo()],
            ['ai_create_agent', () => this.testAICreateAgent()],
            ['ai_login', () => this.testAILogin()],
            ['ai_logout', () => this.testAILogout()],
            ['ai_move', () => this.testAIMove()],
            ['ai_stop', () => this.testAIStop()],
            ['ai_chat', () => this.testAIChat()],
            ['ai_observe', () => this.testAIObserve()],
            ['ai_attack', () => this.testAIAttack()],
            ['ai_equip', () => this.testAIEquip()],
            ['ai_collect', () => this.testAICollect()],
            ['ai_craft', () => this.testAICraft()],
            ['ai_enter', () => this.testAIEnter()]
        ];

        // Run each test
        for (const [testName, testFunc] of testMethods) {
            try {
                await this.suite.runTest(testName, testFunc);
            } catch (error) {
                this.suite.logger.error(`Error running test ${testName}: ${error.message}`);
            }
        }

        this.suite.endTime = Date.now();

        // Cleanup
        if (this.testTokens.length > 0) {
            await cleanupTestAgents(this.config, this.testTokens);
        }

        // Generate report
        this.suite.saveReport();
        this.suite.printSummary();
    }
}

async function main() {
    try {
        const config = new TestConfig();
        const tests = new ServerAPITests(config);
        await tests.runAllTests();
    } catch (error) {
        console.error(`Error: ${error.message}`);
        process.exit(1);
    }
}

if (require.main === module) {
    main();
}

module.exports = ServerAPITests; 